internal/llrp/behavior.go

```diff
@@ -0,0 +1,993 @@
+			if tag.PeakRSSI == nil {
+				tag.PeakRSSI = new(PeakRSSI)
+				*tag.PeakRSSI = *d.lastData.PeakRSSI
+			} else {
+				*d.lastData.PeakRSSI = *tag.PeakRSSI
+			}
+		}
+
+		if d.report.EnableChannelIndex {
+			if tag.ChannelIndex == nil {
+				tag.ChannelIndex = new(ChannelIndex)
+				*tag.ChannelIndex = *d.lastData.ChannelIndex
+			} else {
+				*d.lastData.ChannelIndex = *tag.ChannelIndex
+			}
+		}
+
+		if d.report.EnableFirstSeenTimestamp {
+			if tag.FirstSeenUTC == nil {
+				tag.FirstSeenUTC = new(FirstSeenUTC)
+				*tag.FirstSeenUTC = *d.lastData.FirstSeenUTC
+			} else {
+				*d.lastData.FirstSeenUTC = *tag.FirstSeenUTC
+			}
+		}
+
+		if d.report.EnableLastSeenTimestamp {
+			if tag.LastSeenUTC == nil {
+				tag.LastSeenUTC = new(LastSeenUTC)
+				*tag.LastSeenUTC = *d.lastData.LastSeenUTC
+			} else {
+				*d.lastData.LastSeenUTC = *tag.LastSeenUTC
+			}
+		}
+
+		if d.report.EnableTagSeenCount {
+			if tag.TagSeenCount == nil {
+				tag.TagSeenCount = new(TagSeenCount)
+				*tag.TagSeenCount = *d.lastData.TagSeenCount
+			} else {
+				*d.lastData.TagSeenCount = *tag.TagSeenCount
+			}
+		}
+	}
+}
+
+// Transmit returns a legal llrp.RFTransmitter value.
+func (d *BasicDevice) Transmit(b Behavior) (*RFTransmitter, error) {
+	// First, find the highest power at or below the Target.
+	pwrIdx, pwr := d.findPower(b.Power.Max)
+	if pwr > b.Power.Max {
+		return nil, errors.Wrapf(ErrUnsatisfiable,
+			"target power (%.2f dBm) is lower than the lowest supported (%.2f dBm)",
+			float32(b.Power.Max)/100.0, float32(pwr)/100.0)
+	}
+
+	// In hopping regulatory regions, we assume the power is legal for all frequencies.
+	// This also assumes the first HopTableID is acceptable/worth using,
+	// which may or may not be true.
+	if d.allowsHop {
+		return &RFTransmitter{
+			HopTableID:         uint16(d.freqInfo.FrequencyHopTables[0].HopTableID),
+			TransmitPowerIndex: pwrIdx,
+		}, nil
+	}
+
+	// For non-hopping regions, we need to find a frequency that permits this power level.
+	for _, permitted := range b.Frequencies {
+		for i, f := range d.freqInfo.FixedFrequencyTable.Frequencies {
+			if permitted == f {
+				// +1 because channel indices are 1-based in LLRP
+				return &RFTransmitter{ChannelIndex: uint16(i + 1)}, nil
+			}
+		}
+	}
+
+	return nil, errors.Wrapf(ErrUnsatisfiable,
+		"no frequency permits the desired power level (%.2f dBm)",
+		float32(b.Power.Max)/100.0)
+}
+
+// findPower returns the device's best match to a given power level,
+// suitable for use as the RFTransmitter index in AntennaConfigurations.
+//
+// The returned power level and its respective index
+// is the highest supported power level less than or equal to the target;
+// if the target is less than even the lowest supported power level,
+// then this returns the lowest power level and its respective index,
+// so you should check the value upon return if a higher level is never suitable.
+//
+// This panics if there is not at least one power value.
```

Comment:
Should not `panic` . Should handle the error gracefully.

Reply:
There is no graceful way to handle this error, as it's not meant as part of the public API. Instead, the input is checked when constructing the `BasicDevice` initially (and an error is returned then). If this code get called _without_ a power value, it means that someone changed the code in a way that subvert assumptions this relies upon. It is a programmer error akin to indexing past the end of an array (which, incidentally, is the source of the `panic`). This is also the reason that the code assumes the list is pre-sorted.

Reply:
Please expand comment with this why information. I.e. `If this code get called without a power value, it means that someone changed the code in a way that subvert assumptions this relies upon.`